test/jdk/java/net/Socket/Timeouts.java

```diff
@@ -95,141 +96,182 @@ public void testTimedConnect3() throws IOException {
         withConnection((s1, s2) -> {
             s2.setSoTimeout(2000);
-            expectThrows(SocketTimeoutException.class, () -> s2.getInputStream().read());
+            assertThrows(SocketTimeoutException.class, () -> s2.getInputStream().read());
             s1.getOutputStream().write(99);
             s2.setSoTimeout(0);
             int b = s2.getInputStream().read();
-            assertTrue(b == 99);
+            assertEquals(99, b);
         });
     }
 
     /**
      * Test untimed read that succeeds after a previous read has timed out and
-     * after a short delay
+     * after a short delay.
      */
-    public void testTimedRead7() throws IOException {
+    @Test
+    void testTimedRead7() throws IOException {
         withConnection((s1, s2) -> {
             s2.setSoTimeout(2000);
-            expectThrows(SocketTimeoutException.class, () -> s2.getInputStream().read());
+            assertThrows(SocketTimeoutException.class, () -> s2.getInputStream().read());
             scheduleWrite(s1.getOutputStream(), 99, 2000);
             s2.setSoTimeout(0);
             int b = s2.getInputStream().read();
-            assertTrue(b == 99);
+            assertEquals(99, b);
         });
     }
 
     /**
-     * Test async close of timed read
+     * Test async close of timed read.
      */
-    public void testTimedRead8() throws IOException {
+    @Test
+    void testTimedRead8() throws IOException {
         withConnection((s1, s2) -> {
             s2.setSoTimeout(30*1000);
             scheduleClose(s2, 2000);
-            expectThrows(SocketException.class, () -> s2.getInputStream().read());
+            assertThrows(SocketException.class, () -> s2.getInputStream().read());
         });
     }
 
     /**
-     * Test read with a timeout of Integer.MAX_VALUE
+     * Test read with a timeout of Integer.MAX_VALUE.
      */
-    public void testTimedRead9() throws IOException {
+    @Test
+    void testTimedRead9() throws IOException {
         withConnection((s1, s2) -> {
             scheduleWrite(s1.getOutputStream(), 99, 2000);
             s2.setSoTimeout(Integer.MAX_VALUE);
             int b = s2.getInputStream().read();
-            assertTrue(b == 99);
+            assertEquals(99, b);
+        });
+    }
+
+    /**
+     * Test concurrent reading with a timeout of 2s. The read method should throw
+     * SocketTimeoutException after 2s, not after 2s, 4s, 6s, 8s, ...
+     */
+    @Test
+    void testTimedRead10() throws Exception {
+        var futures = new ArrayList<Future<?>>();
+        withConnection((_, s) -> {
+            s.setSoTimeout(2000);
+            Callable<?> timedReadTask = () -> {
+                long startMillis = millisTime();
+                assertThrows(SocketTimeoutException.class,
+                        () -> s.getInputStream().read());
+                int timeout = s.getSoTimeout();
+                checkDuration(startMillis, timeout-100, timeout+20_000);
```

Comment:
Maybe it would be worth noting here that if the bug isn't fixed, the last thread to call read() would have to wait for 200 seconds, which largely exceeds 22s?